apiserver/common/networkingcommon/networkconfigapi.go

```diff
@@ -117,29 +129,229 @@ func (api *NetworkConfigAPI) getMachineForSettingNetworkConfig(machineTag string
+}
 
-	logger.Debugf("setting devices: %+v", devicesArgs)
-	if err := m.SetParentLinkLayerDevicesBeforeTheirChildren(devicesArgs); err != nil {
-		return errors.Trace(err)
+func (o *updateMachineLinkLayerOp) processExistingDevice(dev LinkLayerDevice) ([]txn.Op, error) {
+	incomingDev := o.MatchingIncoming(dev)
+
+	if incomingDev == nil {
+		ops, err := o.processExistingDeviceNotObserved(dev)
+		return ops, errors.Trace(err)
 	}
 
-	logger.Debugf("setting addresses: %+v", devicesAddrs)
-	if err := m.SetDevicesAddresses(devicesAddrs...); err != nil {
-		return errors.Trace(err)
+	ops := dev.UpdateOps(networkDeviceToStateArgs(*incomingDev))
+
+	incomingAddrs := o.MatchingIncomingAddrs(dev.MACAddress())
+
+	for _, addr := range o.DeviceAddresses(dev) {
+		existingAddrOps, err := o.processExistingDeviceAddress(dev, addr, incomingAddrs)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		ops = append(ops, existingAddrOps...)
 	}
 
-	logger.Debugf("updated machine %q network config", m.Id())
-	return nil
+	newAddrOps, err := o.processExistingDeviceNewAddresses(dev, incomingAddrs)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	o.MarkDevProcessed(dev.MACAddress())
+	return append(ops, newAddrOps...), nil
+}
+
+// processExistingDeviceNotObserved returns transaction operations for
+// processing a device we have in state, but that the machine agent no
+// longer observes locally.
+// The device itself is not marked for deletion now, but for later processing
+// to ensure it is not a parent of other observed devices.
+func (o *updateMachineLinkLayerOp) processExistingDeviceNotObserved(dev LinkLayerDevice) ([]txn.Op, error) {
+	addrs := o.DeviceAddresses(dev)
+
+	var ops []txn.Op
+	var removing int
+	for _, addr := range addrs {
+		// If the machine is the authority for this address,
+		// we can delete it; otherwise leave it alone.
+		if addr.Origin() == network.OriginMachine {
+			logger.Debugf("removing address %q from device %q", addr.Value(), dev.Name())
+			ops = append(ops, addr.RemoveOps()...)
+			removing++
+		}
+	}
+
+	// If the device is having all of its addresses removed and is not under
+	// the authority of the provider, add it as a candidate for removal.
+	// If the device has been relinquished by the provider, the instance-poller
+	// will have removed the provider ID - see the instance-poller API facade
+	// logic.
+	if removing == len(addrs) && dev.ProviderID() == "" {
+		o.removalCandidates = append(o.removalCandidates, dev)
+	}
+
+	return ops, nil
+}
+
+func (o *updateMachineLinkLayerOp) processExistingDeviceAddress(
+	dev LinkLayerDevice, addr LinkLayerAddress, incomingAddrs []state.LinkLayerDeviceAddress,
+) ([]txn.Op, error) {
+	addrValue := addr.Value()
+
+	// If one of the incoming addresses matches the existing one,
+	// update it.
+	for _, incomingAddr := range incomingAddrs {
+		if strings.HasPrefix(incomingAddr.CIDRAddress, addrValue) &&
+			!o.IsAddrProcessed(dev.MACAddress(), incomingAddr.CIDRAddress) {
```

Comment:
Should this match logic be extracted into a helper method on `updateMachineLinkLayerOp`?